flocker/node/agents/gce.py

```diff
@@ -402,19 +413,48 @@ def allocation_unit(self):
 
     def list_volumes(self):
         # TODO(mewert) Walk the pages.
-        result = self._compute.disks().list(project=self._project,
-                                            zone=self._zone).execute()
-        return list(
-            BlockDeviceVolume(
-                blockdevice_id=unicode(disk['name']),
-                size=int(GiB(int(disk['sizeGb'])).to_Byte()),
-                attached_to=_extract_attached_to(disk),
-                dataset_id=_blockdevice_id_to_dataset_id(disk['name'])
+        with start_action(
+                action_type=u"flocker:node:agents:gce:list_volumes",
+        ) as action:
+            result = self._compute.disks().list(project=self._project,
+                                                zone=self._zone).execute()
+
+            # 'description' will not even be in the dictionary if no
+            # description was specified.
+            def disk_in_cluster(disk):
+                return (
+                    disk['name'].startswith(_PREFIX) and
+                    (disk.get('description') ==
+                     self._disk_resource_description())
+                )
+
+            Message.log(
+                message_type=u'flocker:node:agents:gce:list_volumes:ignored',
+                ignored_volumes=list({'name': disk['name'],
+                                      'description': disk.get('description')}
+                                     for disk in result['items']
+                                     if not disk_in_cluster(disk))
```

Comment:
Given that disk_in_cluster is a pure function I don't think it's that wasteful, and think that's slightly worse for memory pressure if the size of the lists are approximately equal. (This list can be deleted here and the same memory can be re-used for the next list.)

With lists this small though, readability/maintainability is probably preferred over complexity/size analysis.

I have a vague idea that I've written fewer bugs with this functional-style formation rather than the iterate through and add to one list or the other, but no particular reason to prefer one over the other. Given that you prefer the other I'll go ahead with that formation.